streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -747,23 +747,18 @@ private void runLoop() {
                 if (assignmentErrorCode.get() == AssignorError.VERSION_PROBING.code()) {
                     log.info("Version probing detected. Triggering new rebalance.");
                     assignmentErrorCode.set(AssignorError.NONE.code());
-                    enforceRebalance();
+                    consumer.enforceRebalance();
                 }
             } catch (final TaskMigratedException ignoreAndRejoinGroup) {
                 log.warn("Detected task {} that got migrated to another thread. " +
                              "This implies that this thread missed a rebalance and dropped out of the consumer group. " +
                              "Will try to rejoin the consumer group.", ignoreAndRejoinGroup.migratedTaskId());
 
-                enforceRebalance();
+                consumer.enforceRebalance();
             }
         }
     }
 
-    private void enforceRebalance() {
-        consumer.unsubscribe();
-        subscribeConsumer();
```

Comment:
Previously in `unsubscribe` we do make sure all tasks are closed before re-subscribing, now we are relying on the consumer's response to let us know which (subset of) tasks are to be closed -- so although the task-migrated is triggered by fenced, if the heartbeat thread has not got the error code and reset the group info we would not call `onPartitionsLost` before re-joining the group, in that case would we possibly have more than one instance claiming to own some same partitions? Worth checking the consumer.assign function to see if the code already handles this case.

Generally speaking I do not know if it would expose any edge cases or not. Better run our system tests a couple of times to make sure?

Reply:
That's a fair point, I think we actually do want to use the old method of unsubscribing in the specific case of a `TaskMigratedException`, and this new API only for version probing, KIP-441, etc.

I don't think it would be _incorrect_ to do this, just inefficient -- if the heartbeat thread hasn't reset the group by the next poll call, the consumer will call `onPartitionsRevoked` (eager) or nothing (cooperative) and send a `JoinGroup`. On the response it will realize it dropped out and then invoke `onPartitionsLost`.

Reply:
@guozhangwang Actually, looking at `onLeavePrepare` it seems if the heartbeat thread hasn't reset the group then we won't call `onPartitionsLost` even on `unsubscribe` 
You partially addressed this [here](https://github.com/apache/kafka/pull/7997/files#r374990440), but what happens if we aren't in a rebalance yet? I suppose we could actually combine the old method and the new by calling `enforceRebalance` and _then_ `unsubscribe` but that feels a bit hacky. WDYT? 
Looking at this again I'm not sure I understand/remember why it's correct to invoke `onPartitionsLost` instead of `onPartitionsRevoked` if a rebalance is in progress during `unsubscribe` -- can you explain that a bit further?

Reply:
There's always a race condition because the fencing error could be notified from the brokers in two ways: 1) heartbeat, 2) commit (including txnal send-offsets). Either one could be ahead of the other in getting the error code.

When we get task-migrated it means 2) got the error, at that time 1) may not get it yet so the local state may still be "stable" and generation not reset yet. In that case here we would `revokeAll` still and committing which is doom to fail again, otherwise we would `lostAll` (note the PR you referenced just make sure that "otherwise" case happens). In my other PR I also realized this from some system test failures and did this change to always "handleLostAll" before re-subscribing:

https://github.com/apache/kafka/pull/8058/files#diff-045aeaddb4232a85a8560186b4901e69R763

In some sense this is to be conservative and make sure when re-subscribing we do not encode any "owned partitions".

Reply:
I'm also a bit concerned with this change, as we are calling `onJoinPrepare` before `onLeavePrepare` which changes the existing sequence for no specific reason.

Reply:
@guozhangwang right, in that case isn't the current behavior also at risk of this race condition? If we hit 2) even `unsubscribe`/`onLeavePrepare` will only call `onPartitionsRevoked` and not `onPartitionsLost` 
Seems we should try to fix this on the consumer client level for 2.6/trunk, as it's invoking the wrong callback. I'm not sure it's possible to detect this smoothly however, in which case we can handle the `TaskMigratedException` case by first calling the new `enforceRebalance` API and then calling `unsubscribe` -- since you fixed `onLeavePrepare` to now call `onPartitionsLost` in one of your cleanup PRs.

This still leaves us exposed in 2.5...see comment [here](https://github.com/apache/kafka/pull/8145/files#r382782577)

Reply:
@abbccdda we actually do want to call `onJoinPrepare` before `onLeavePrepare` for the specific reason that we don't want to give up all our partitions unnecessarily -- of course, in the `TaskMigratedException` case we do still want to lose all owned partitions

Reply:
The current condition is only `generation() == Generation.NO_GENERATION || rebalanceInProgress()` and `enforceRebalance` only sets `rebalanceNeeded` to false, so after that calling `unsubscribe` sill does not guarantee we trigger `onPartitionsLost` instead of `onPartitionsRevoked` right?

Reply:
Right, yeah, I was confusing `MemberState.REBALANCING` with `needsRejoin`. Calling `enforceRebalance` first wouldn't help -- I guess my preference would be to fix this in the consumer client somehow, ie make sure it invokes the correct callback during `unsubscribe`. For example if we get `TaskMigratedException` due to a failed commit, the consumer will reset the generation itself. But I'm not sure we can rely on it resetting the generation for any arbitrary cause of `TaskMigratedException` -- which can happen for reasons outside of consumer calls failing IIUC
If we can't fix in the consumer client, then I would agree we should catch `TaskMigratedException` and call `handleLostAll` followed by `unsubscribe` (and resubscribe).

Reply:
I guess my _real_ preference would be to consider adding a second new Consumer API as part of this KIP, which acts similar to calling `unsubscribe` -> `resubscribe` but always invokes `onPartitionsLost`. Something like `rejoinGroupAsZombie` or `rejoinAfterDroppingOut` to force the consumer to acknowledge that it's dropped out and lost all its partitions. 
Alternatively we could add a Consumer API that basically just calls `resetGeneration`, and then leave it up to the user to call `unsubscribe` or `enforceRebalance` themselves. But, not sure if this is exposing too much through the Consumer interface

Reply:
Even if ConsumerCoordinator already reset generation on all occasions when throwing the consumer fenced exception that is still not safe, sine we could still be fenced on the producer side.

Personally I'm okay with either calling `handleAllLost` before unsubscribe, or adding a boolean parameter as in `enforceRebalance(boolean resetGeneration)`, seems @ableegoldman is preferring the latter here so I'm giving it to you :) _(feeling we are making the new API more and more advanced-usage only now)_

Reply:
Yeah, I'd rather handle it within Streams or add a new API altogether than start adding semi-orthogonal parameters to this API. Since we _can_ handle this in a reasonable way within Streams (for now) I'm fine with doing so and we can always revisit the issue later